move goals service to separate service init commit

mkhomishchak opened the pull request.

mkhomishchak committed:
move goals service to separate service init commit

mkhomishchak committed:
set userId on crypto table update request

mkhomishchak committed:
add transactions with creating new records for table

before now users could create update transactions only for the records that already existed in the table
with given new endpoint we will be able to create new records by single transactions, without updating whole goals table

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/config/GoalServiceConfig.java:
```diff
@@ -0,0 +1,14 @@
+public class GoalServiceConfig {
+
+    @Bean("goalServiceRestTemplate")
+    public RestTemplate getRestTemplate() {
```
Why do you need to qualify it ?

mkhomishchak replied:
isn't it better to always qualify beans in order to avoid any potential mismatches?

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/controller/GoalsController.java:
```diff
@@ -0,0 +1,73 @@
+
+    @GetMapping("/crypto-tables")
+    public ResponseEntity<CryptoGoalsTable> getCryptoGoalsTable(@RequestHeader("UserId") Long userId) {
+        return new ResponseEntity<>(goalsService.getCryptoGoalsTable(userId), HttpStatus.OK);
```
```suggestion
        return ResponseEntity.ok(goalsService.getCryptoGoalsTable(userId));
```

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/repository/SelfGoalRepository.java:
```diff
@@ -0,0 +1,15 @@
+
+    List<SelfGoal> findAllByUserId(Long userId);
+
+    @Query(value = "SELECT * FROM self_goals sf WHERE NOW() > sf.end_date AND sf.is_closed = false", nativeQuery = true)
```
Why don't use query builder by method naming ? E.g `findAllByEndDateLessThanAndIsClosed(Instant endDate, boolean isClosed)`

mkhomishchak replied:
isn't my one better? I should not pass any additional fields for this method, and for so far I did have just one single usage of this method, so not sure this change will somehow improve it

mkhomishchak committed:
updated call to balance service for transactions history deposit amount

the fix was to add userId custom header as we do not have any token to go through auth service, and no creds to create token
and in as war as this will be a public API without token validation and so userId exctraction we are passing it mannually
note it is a quick fix, it will be replaced with event driving architecture soon

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/controller/GoalsController.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @GetMapping("/self-goals")
+    public ResponseEntity<List<SelfGoal>> getSelfGoals(@RequestHeader("UserId") Long userId) {
```
`"UserId"` can be moved to some static variable as it's used within application in may places

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/model/transaction/CreateNewRecordTransaction.java:
```diff
@@ -0,0 +1,15 @@
+
+public record CreateNewRecordTransaction(String name, BigDecimal quantity, BigDecimal price, BigDecimal goalQuantity) {
+
+    public CreateNewRecordTransaction {
```
Can be converted to `BigDecimal.ZERO.compareTo(...) > 0`. But it's not critical

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/model/transaction/CreateNewRecordTransaction.java:
```diff
@@ -0,0 +1,15 @@
+        if (quantity != null && quantity.compareTo(BigDecimal.ZERO) < 0) {
+            throw new IllegalArgumentException("Quantity cannot be less than zero.");
+        }
+        if (goalQuantity != null && goalQuantity.compareTo(BigDecimal.ZERO) < 0) {
```
Same as above

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/service/GoalsServiceImpl.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Override
+    public CryptoGoalsTable updateCryptoGoalsTable(CryptoGoalTableTransaction transaction, long tableId) {
```
```suggestion
    @Transactional
    public CryptoGoalsTable updateCryptoGoalsTable(CryptoGoalTableTransaction transaction, long tableId) {
```

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/service/GoalsServiceImpl.java:
```diff
@@ -0,0 +1,243 @@
+    public CryptoGoalsTable getCryptoGoalsTable(Long userId) {
+        CryptoGoalsTable table  = cryptoGoalsTableRepository.findByUserId(userId).orElseThrow(
+                () -> new GoalsTableNotFoundException(String.format("could not find goals table for userId: %d", userId)));
+        table.getTableRecords().forEach(this::setPostQuantityValues);
```
`n+1` problem as `tableRecords` are loaded lazily

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/service/scheduled/GoalsScheduledService.java:
```diff
@@ -0,0 +1,52 @@
+    @Scheduled(cron = "0 0 0 * * ?")
+    public void doAtTheStartOfTheDay() {
+        List<SelfGoal> overdueGoals = goalsService.getAllOverdueGoals();
+        closeOverdueAndCreateNewGoals(overdueGoals);
```
`n+1`

DendeberiaOleksandr commented on src/main/java/com/khomishchak/goalsservice/service/scheduled/GoalsScheduledService.java:
```diff
@@ -0,0 +1,52 @@
+        overdueGoals.forEach(overdueGoal -> {
+            closeOverdueGoal(overdueGoal);
+            SelfGoal newGoal = createNewFromOverdueGoal(overdueGoal);
+            goalsService.saveAll(List.of(overdueGoal, newGoal));
```
Could you please:
1) fix n+1 problem
2) change the algorithm if it's possible: first of all collect all `overfueGoals` and then save all by single `saveAll()` not by invoking `saveAll(...)` for every `overdueGoal`

mkhomishchak committed:
applied suggestions

Max-Homishchak committed:
small refactor
